### PR TITLE
AirfRANS: Depth scaling 5L/256d with golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The AirfRANS baseline (0.007264) was achieved at 4L/256d with the golden config (gc=1.0, WD=1e-2, T_max=5). Scaling from 3L→4L gave a 22.3% improvement. We hypothesize that 5L/256d will capture more complex flow patterns with additional depth, potentially closing the gap to the external target (0.0043). The golden config (WD+gc) stabilized 4L training — it should also stabilize 5L. Key risk: slower epochs may reduce total epoch count within the 180-min budget.

## Instructions
Train on AirfRANS with 5L/256d architecture + full golden config:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent armin \
  --wandb-name "armin/airfrans-5L256d-golden" \
  --wandb-group "airfrans-depth-scaling"
```

CRITICAL FLAGS: --no-use-ema (EMA bug), --epochs 999 (default is 2), SENPAI_MAX_EPOCHS=9999 (default 50 cap), SENPAI_TIMEOUT_MINUTES=180.

Report `val_primary/surface_mse` (lower is better) and note which epoch was best, total epochs completed, and time per epoch.

## Baseline
Current best: **0.007264** at epoch 40 (4L/256d, gc=1.0, WD=1e-2, T_max=5, 50 epochs in 61 min)
PR: #2727 | W&B: ruurxdqs
External target: 0.0043 (1.7x gap)

Reproduce: `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT armin:

## Results

### Experiment: AirfRANS 5L/256d with golden config

| Metric | Baseline (PR #2727, 4L/256d) | This Experiment (5L/256d) | Delta |
|--------|------------------------------|---------------------------|-------|
| val_primary/surface_mse | **0.007264** | **0.005206** | **-28.3%** |
| full_val/volume_mse | — | 0.0429 | — |
| full_val/surface_mse_p | — | 0.0208 | — |
| External target | 0.0043 | 0.005206 | 21.1% gap remaining |

**Best epoch**: 56 (of 83 completed before late-training divergence)

### Training Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent armin \
  --wandb-name "armin/airfrans-5L256d-golden" \
  --wandb-group "airfrans-depth-scaling"
```

- **W&B run**: `zc6sryys`
- **Peak memory**: ~8.5 GB VRAM (GPU 0)
- **Epochs completed**: 83 (still running to timeout, but diverged at epoch 72)
- **Time per epoch**: ~1.5 min (vs ~1.2 min at 4L/256d — 25% slower as expected)

### What happened — Strong positive result with late instability

**This is a positive result.** 5L/256d achieves a 28.3% improvement over the 4L/256d baseline, closing the gap to the external target (0.0043) from 1.7x to 1.2x.

**Two-phase training behavior:**

**Phase 1: Rapid convergence (epochs 1-71)**
- Trough envelope descended consistently: 0.220→0.180→0.057→0.029→0.012→0.0074→0.0067→0.0052
- Phase transition occurred at epoch 16-17 (val halved from 0.131→0.072→0.057)
- Beat 4L/256d baseline (0.007264) for the first time at epoch 52 (val=0.00671)
- Reached best val=0.005206 at epoch 56
- Grad norms were well-behaved: 10→19 (early), then decreasing to 7.9 at best epoch
- Grad clip events decreased from 359/360 to 285/360, indicating model stabilization

**Phase 2: Catastrophic late-training instability (epochs 72-83+)**
- Gradient norms exploded: 284→641→193→464→574→535→780→404→1433→900→3313→3152
- Val spiked to 0.16→0.31→0.39 with no recovery
- All 360 batches hitting grad clip ceiling (360/360)
- The model entered an unrecoverable divergent state

**Why it worked (and then didn't):**

The additional 5th layer provides ~25% more parameters and a deeper feature hierarchy, allowing the model to capture more complex flow patterns (boundary layer interactions, wake structures). With the golden config's WD=1e-2 and gc=1.0, training was stable and productive for 71 epochs.

The late instability suggests that at 5L depth, the model eventually reaches a region of the loss landscape where gradient paths through 5 layers amplify perturbations beyond what gc=1.0 can contain. This is consistent with the known depth-stability tradeoff: deeper models have higher Lipschitz constants, making them more susceptible to gradient explosion in certain loss landscape regions. Notably, this did NOT happen at 4L/256d (which was stable for 50 epochs and still improving), confirming that 5L pushes the depth-stability boundary.

### Key takeaways

1. **Depth scaling from 4L→5L works**: 28.3% improvement, confirming the hypothesis
2. **Diminishing returns**: 3L→4L gave 22.3% improvement; 4L→5L gave 28.3%. Returns are actually NOT diminishing — the 5th layer provided more benefit, likely because the golden config's regularization enables full exploitation of extra capacity
3. **Instability is the limiting factor**: The model reached its best at epoch 56 then diverged at epoch 72. A stabilization mechanism is needed for long 5L training
4. **External target gap closing**: Now at 1.2x (was 1.7x at 4L)

### Suggested follow-ups

1. **5L/256d with gc=0.5**: Tighter grad clipping may prevent the late-training instability while preserving the convergence benefit. The instability occurred when grad norms started exceeding gc=1.0 by 100-1000x — a tighter clip would catch it earlier.
2. **5L/256d with lr=5e-4 + gc=1.0**: Lower LR reduces gradient magnitudes and may avoid the instability region entirely. PR #2719 showed lr=5e-4 is very effective at 3L/192d.
3. **5L/256d with checkpoint at epoch 56**: If checkpointing were enabled, we could resume from the best model and try fine-tuning with a lower LR to squeeze more performance.
4. **6L/256d**: Given 5L showed non-diminishing returns, 6L is worth exploring — but will need aggressive stabilization (gc=0.5 + lr=5e-4) to handle even deeper gradient paths.
5. **5L/256d with warmup**: A learning rate warmup period might help the deeper model establish stable gradient flow patterns before ramping up to full LR.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2799 because ADVISOR: Closing — the AirfRANS baseline was just updated to 0.003904 (PR #2755 shoya, same golden config with SENPAI_MAX_EPOCHS=9999 + 180-min). Your 5L/256d result (0.005206 at ep56) is 33% worse than the new baseline. However, the finding is valuable: 5L converges 3x faster per epoch (0.005206 at ep56 vs 4L needing ep201 for 0.003904), but diverges early (ep72 vs ep208). Depth scaling works — just needs stability fixes. Excellent analysis and suggestions. Assigning you a 5L stability follow-up next..
